java/src/Ice/src/main/java/com/zeroc/IceInternal/HashUtil.java

```diff
@@ -6,74 +6,78 @@
 
 public final class HashUtil {
   public static int hashAdd(int hashCode, boolean value) {
-    return ((hashCode << 5) + hashCode) ^ (value ? 0 : 1);
+    return shift(hashCode) ^ (value ? 0 : 1);
   }
 
-  public static int hashAdd(int hashCode, short value) {
-    return ((hashCode << 5) + hashCode) ^ (int) (2654435761l * value);
+  public static int hashAdd(int hashCode, byte value) {
+    return shift(hashCode) ^ (int) (2654435761l * value);
   }
 
-  public static int hashAdd(int hashCode, byte value) {
-    return ((hashCode << 5) + hashCode) ^ (int) (2654435761l * value);
+  public static int hashAdd(int hashCode, short value) {
+    return shift(hashCode) ^ (int) (2654435761l * value);
   }
 
   public static int hashAdd(int hashCode, int value) {
-    return ((hashCode << 5) + hashCode) ^ (int) (2654435761l * value);
+    return shift(hashCode) ^ (int) (2654435761l * value);
   }
 
   public static int hashAdd(int hashCode, long value) {
-    return ((hashCode << 5) + hashCode) ^ (int) (value ^ (value >>> 32));
+    return shift(hashCode) ^ (int) (value ^ (value >>> 32));
   }
 
   public static int hashAdd(int hashCode, float value) {
-    return ((hashCode << 5) + hashCode) ^ Float.floatToIntBits(value);
+    return shift(hashCode) ^ Float.floatToIntBits(value);
   }
 
   public static int hashAdd(int hashCode, double value) {
     long v = Double.doubleToLongBits(value);
-    return ((hashCode << 5) + hashCode) ^ (int) (v ^ (v >>> 32));
+    return shift(hashCode) ^ (int) (v ^ (v >>> 32));
   }
 
   public static int hashAdd(int hashCode, java.lang.Object value) {
```

Comment:
I find this Object error-prone. Do we really need it?

Keep in mind this HashUtil isn't/shouldn't be general-purpose Hash helpers. And it would be nicer to drop the public visibility.

Reply:
What do you find error-prone about this function?
And yes, we need it. Obviously we never pass in a literal `Object`, but we pass in other stuff like `Identity`, `EncodingImage`, `Map<...>`, any non-primitive type

Reply:
As for the public visibility, we're actually pretty good about using this code only internally (for proxies and endpoints), but there is one place where we use it publicly: the generated `hashCode` function for `struct`s.

Reply:
For error-prone, my concern is we can accidentally pass a class without a custom hashCode implementation... and then we'll hash the reference. Maybe that's fine?

Reply:
There really isn't a good solution here.
Since, there is no `Hashable` interface we could use as a type-bound; `Object` is inherently `Hashable` in Java.

But also: I think it's fine. If a type doesn't define a custom hash, then it's fine to use the 'default' reference.